### PR TITLE
Feat: SQL performance

### PR DESCRIPTION
**Pull request by lucas-zimerman**
This PR Adds automatic spans to any code that interacts with 

https://www.nuget.org/packages/Microsoft.Data.SqlClient/
and 
https://www.nuget.org/packages/System.Data.SqlClient.

The following Spans are created:
* db.connection: measures the connection lifespan.
* db.query: measures the time required to execute the query.

### CODE REVIEW

File: src/Sentry.Extensions.Logging/SentryDiagnosticSubscriber.cs [hunk 1]
```diff
@@ -30,6 +30,11 @@ public void OnNext(DiagnosticListener listener)
                 _efInterceptor = new(_hub, _options);
                 listener.Subscribe(_efInterceptor);
             }
+            else if(listener.Name == "SqlClientDiagnosticListener")
```
Comment (bruno-garcia) [inline src/Sentry.Extensions.Logging/SentryDiagnosticSubscriber.cs:33]:
```suggestion
            else if (listener.Name == "SqlClientDiagnosticListener")
```

File: src/Sentry.Extensions.Logging/SentrySqlListener.cs [hunk 2]
```diff
@@ -0,0 +1,182 @@
...
+        private SentryOptions _options { get; }
+        public SentrySqlListener(IHub hub, SentryOptions options)
+        {
+            _hub = hub;
+            _options = options;
+        }
+
+        private void AddSpan(SentrySqlSpanType type, string operation, string? description, Guid operationId, Guid? connectionId = null)
+        {
+            _hub.WithScope(scope =>
```
Comment (bruno-garcia) [inline src/Sentry.Extensions.Logging/SentrySqlListener.cs:50]:
```suggestion
            _hub.ConfigureScope(scope =>
```

File: src/Sentry.Extensions.Logging/SentrySqlListener.cs [hunk 3]
```diff
@@ -0,0 +1,182 @@
...
+
+        private ISpan? TryGetConnectionSpan(Scope scope, Guid connectionId)
+            => scope.Transaction?.Spans.FirstOrDefault(span => TryGetKey(span.Extra, ConnectionExtraKey) is Guid id && id == connectionId);
+
+        private ISpan? TryGetQuerySpan(Scope scope, Guid operationId)
+            => scope.Transaction?.Spans.FirstOrDefault(span => TryGetKey(span.Extra, OperationExtraKey) is Guid id && id == operationId);
+
+        private void UpdateConnectionSpan(Guid operationId, Guid connectionId)
+        {
+            _hub.WithScope(scope =>
```
Comment (bruno-garcia) [inline src/Sentry.Extensions.Logging/SentrySqlListener.cs:101]:
```suggestion
            _hub.ConfigureScope(scope =>
```

File: src/Sentry.Extensions.Logging/SentrySqlListener.cs [hunk 4]
```diff
@@ -0,0 +1,182 @@
...
+                    var span = parentSpan.StartChild(operation, description);
+                    span.SetExtra(OperationExtraKey, operationId);
+                }
+            });
+        }
+
+        private ISpan? GetSpan(SentrySqlSpanType type, Guid? operationId = null, Guid? connectionId = null)
+        {
+            ISpan? span = null;
+            _hub.WithScope(scope =>
```
Comment (bruno-garcia) [inline src/Sentry.Extensions.Logging/SentrySqlListener.cs:71]:
```suggestion
            _hub.ConfigureScope(scope =>
```

File: src/Sentry.Extensions.Logging/SentrySqlListener.cs [hunk 5]
```diff
@@ -0,0 +1,182 @@
...
+                }
+                else if (type == SentrySqlSpanType.Connection &&
+                    connectionId is { } id &&
+                    TryGetConnectionSpan(scope, id) is { } connectionSpan)
+                {
+                    span = connectionSpan;
+                }
+                else
+                {
+                    _options.DiagnosticLogger?.LogWarning("Trying to close a span of type {0} with operation id {1}, but it was not found.", type, operationId);
```
Comment (bruno-garcia) [inline src/Sentry.Extensions.Logging/SentrySqlListener.cs:87]:
```suggestion
                    _options.DiagnosticLogger?.LogWarning("Trying to get a span of type {0} with operation id {1}, but it was not found.", type, operationId);
```

File: test/Sentry.Extensions.Logging.Tests/SentrySqlListenerTests.cs [hunk 6]
```diff
@@ -0,0 +1,242 @@
...
+        var x when
+                x == SqlDataBeforeExecuteCommand ||
+                x == SqlMicrosoftBeforeExecuteCommand ||
+                x == SqlDataAfterExecuteCommand ||
+                x == SqlMicrosoftAfterExecuteCommand ||
+                x == SqlDataWriteCommandError ||
+                x == SqlMicrosoftWriteCommandError
+            => (span) => span.Operation == "db.query",
+        _ => throw new NotSupportedException()
+    };
```
Comment (bruno-garcia) [inline test/Sentry.Extensions.Logging.Tests/SentrySqlListenerTests.cs:50]:
```suggestion
		        type switch
		        {
		            var x when
		                    x == SqlDataWriteConnectionOpenBeforeCommand ||
		                    x == SqlMicrosoftWriteConnectionOpenBeforeCommand ||
		                    x == SqlMicrosoftWriteConnectionOpenAfterCommand ||
		                    x == SqlDataWriteConnectionOpenAfterCommand ||
		                    x == SqlMicrosoftWriteConnectionCloseAfterCommand ||
		                    x == SqlDataWriteConnectionCloseAfterCommand
		                => (span) => span.Description is null && span.Operation == "db.connection",
		            var x when
		                    x == SqlDataBeforeExecuteCommand ||
		                    x == SqlMicrosoftBeforeExecuteCommand ||
		                    x == SqlDataAfterExecuteCommand ||
		                    x == SqlMicrosoftAfterExecuteCommand ||
		                    x == SqlDataWriteCommandError ||
		                    x == SqlMicrosoftWriteCommandError
		                => (span) => span.Operation == "db.query",
		            _ => throw new NotSupportedException()
		        };
```

Comment (bruno-garcia) [reply test/Sentry.Extensions.Logging.Tests/SentrySqlListenerTests.cs:50]: Also, please don't name things `x`. Name it the first character of the thing you're dealing with, or the whole name.

**Comment by bruno-garcia** [comment]
As discussed on a call:

* Create a new package `Sentry.Diagnostics.DiagnosticSource`
* New package will: 
  * depends on `System.Diagnostics.DiagnosticSource`
  * target only ns2.0, ns2.1, net461 (newer targets already include S.D.D
* `Sentry` package will include all `../Sentry.Diagnostics.DiagnosticSource/**/*.cs` files when targeting nca3.0 or higher
* Unit tests under `S.D.D` would also be included in `Sentry.Tests`
* Packages that will depend on `Sentry.Diagnostics.DiagnosticSource`:
  * `Sentry.AspNet`
  * `Sentry.AspNetCore` when targeting ns2.0. `#ifdef` to enable the listeners on that TFM

**Comment by lucas-zimerman** [comment]
Closed in favor of #1154

**Pull request closed without merging**